emData/generate_PR.py

```diff
@@ -0,0 +1,103 @@
+    elif any(psword in module for psword in ["L4","L5","L6"]): TProjRegion = "BARREL2S"
+    else: TProjRegion = "DISK"
+    
+    if any(psword in module for psword in ["L1","L2","L3","L4","L5","L6"]): VMProjRegion = "BARREL"
+    else: VMProjRegion = "DISK"
+    
+    return TProjRegion, VMProjRegion
+
+def getLayerAndDisk(module):
+    layer = 0
+    disk = 0
+    if module[3] == 'L': layer = module[4]
+    elif module[3] == 'D': disk = module[4]
+    return layer, disk
+
+def writeHeaderModuleInstance(module):
+    TProjRegion, VMProjRegion = getTProjAndVMProjRegions(module)
+    strHeader = ""
+    strHeader += "void ProjectionRouterTop_" + module[3:] + "(BXType bx,\n"
+    strHeader += "                         const TrackletProjectionMemory<" + TProjRegion + ">*,\n"
+    strHeader += "                         BXType&,\n"
+    strHeader += "                         AllProjectionMemory<" + TProjRegion + ">&,\n"
+    strHeader += "                         VMProjectionMemory<" + VMProjRegion + ">*);\n\n"
+    return strHeader
+
+def writeHeaderPreamble():
+    strPreamble = ""
+    strPreamble += "#ifndef TrackletAlgorithm_ProjectionRouterTop_h\n"
+    strPreamble += "#define TrackletAlgorithm_ProjectionRouterTop_h\n\n"
+    strPreamble += "#include \"ProjectionRouter.h\"\n\n"
+    return strPreamble
+
+def writeHeaderPostamble():
+    strPostamble = ""
+    strPostamble += "#endif\n"
+    return strPostamble
+ 
+def writeSourceModuleInstance(module, wiresFileName):
+    TProjCount, VMProjCount = getNTProjAndNVMProj(module, wiresFileName)
+    TProjRegion, VMProjRegion = getTProjAndVMProjRegions(module)
+    layer, disk = getLayerAndDisk(module)
+    strSource =  ""
+    strSource += "void ProjectionRouterTop_" + module[3:] + "(BXType bx,\n"
+    strSource += "                         const TrackletProjectionMemory<" + TProjRegion + "> projin[" + str(TProjCount) + "],\n" 
+    strSource += "                         BXType& bx_o,\n"
+    strSource += "                         AllProjectionMemory<" + TProjRegion + ">& allprojout,\n"
+    strSource += "                         VMProjectionMemory<" + VMProjRegion + "> vmprojout[" + str(VMProjCount) + "])\n"
+    strSource += "{\n"
+    strSource += "#pragma HLS inline off\n"
+    strSource += "#pragma HLS interface register port=bx_o\n"
+    for i in range(TProjCount):
+        strSource += "#pragma HLS resource variable=projin[" + str(i) + "]->get_mem() latency=2\n"
+    strSource += "  constexpr unsigned int nInMem = " + str(TProjCount) + ";\n"
+    strSource += "  constexpr unsigned int nOutMem = " + str(VMProjCount) + ";\n"
+    strSource += "  constexpr int layer = " + str(layer) + ";\n"
+    strSource += "  constexpr int disk = " + str(disk) + ";\n"
+    strSource += module + ": ProjectionRouter<" + TProjRegion + ", " + VMProjRegion + ", nInMem, nOutMem, layer, disk>\n"
+    strSource += "    (bx, projin, bx_o, allprojout, vmprojout);\n"
+    strSource += "}\n\n"
+    return strSource
```

Comment:
@djcranshaw I guess this function will return nonsense if the module is not in the given wiring? I think either it should return an empty string when `TProjCount` and `VMProjCount` are zero, or you should generate `modules_to_test` from the wiring itself.

Reply:
@aehart Yes, you're right that this function would return nonsense if the module isn't in the given wiring. This is related to the above comment on modules_to_build as well - I had originally envisioned that variable to just hold the names of modules that we know will build top-levels that work, but maybe the right approach is just to build all modules out of the box, and only add the modules that we want to test into the tcl script. And if we do that, there's no need to list the modules individually in their own variable, they can just be built from the wiring file. This approach would mean that we build some top-levels that don't work, but if we're not testing/using them yet anyways maybe that's fine. Also all the modules needed for the summer chain do work.